File: python/csv_reading.py
import csv
from datetime import datetime
from decimal import Decimal
from pathlib import Path

from models import ConsumptionVolume, Duration, ProductionVolume


def parse_timestamp(timestamp_str: str) -> datetime:
    return datetime.fromisoformat(timestamp_str)


def load_consumption(file_path: Path) -> list[ConsumptionVolume]:
    consumption_volumes = []
    with open(file_path, newline='', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        for row in reader:
            volume = ConsumptionVolume(
                consumer=row['consumer'],
                period_start=parse_timestamp(row['timestamp']),
                duration=Duration.HOUR,
                quantity_kwh=Decimal(row['quantity'])
            )
            consumption_volumes.append(volume)
    return consumption_volumes


def load_production(file_path: Path) -> list[ProductionVolume]:
    production_volumes = []
    with open(file_path, newline='', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        for row in reader:
            volume = ProductionVolume(
                producer=row['producer'],
                period_start=parse_timestamp(row['timestamp']),
                duration=Duration.HOUR,
                quantity_kwh=Decimal(row['quantity'])
            )
            production_volumes.append(volume)
    return production_volumes


def load_allocations(file_path: Path) -> None:
    
    with open(file_path, newline='', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        for row in reader:            
            
            producer = row['producer']
            consumer = row['consumer']
            period_start = parse_timestamp(row['timestamp'])
            duration = Duration.HOUR
            quantity_kwh = Decimal(row['quantity'])

            # TODO: return allocations


def get_repo_root(current_dir: Path | None = None) -> Path:
    if current_dir is None:
        current_dir = Path(__file__).parent
    if (current_dir / '.git').is_dir():
        return current_dir
    parent_dir = current_dir.parent
    if parent_dir == current_dir:
        raise Exception("Repository root not found")
    return get_repo_root(parent_dir)

File: typescript/src/csvReading.ts
import * as fs from 'fs';
import * as path from 'path';
import { Temporal } from '@js-temporal/polyfill';
import { parse } from 'csv-parse/sync';
import Decimal from 'decimal.js';
import {
  ConsumptionVolume,
  ProductionVolume,
  Duration,
} from './models';

const TIMEZONE = 'Europe/London';

export function parseTimestamp(timestampStr: string): Temporal.ZonedDateTime {
  const instant = Temporal.Instant.from(timestampStr);
  return instant.toZonedDateTimeISO(TIMEZONE);
}

function stripBom(content: string): string {
  if (content.charCodeAt(0) === 0xfeff) {
    return content.slice(1);
  }
  return content;
}

function readCsvFile(filePath: string): Record<string, string>[] {
  const content = stripBom(fs.readFileSync(filePath, 'utf-8'));
  return parse(content, {
    columns: true,
    skip_empty_lines: true,
  });
}

export function loadConsumption(filePath: string): ConsumptionVolume[] {
  const rows = readCsvFile(filePath);
  return rows.map(
    (row) =>
      new ConsumptionVolume(
        row['consumer'],
        parseTimestamp(row['timestamp']),
        Duration.HOUR,
        new Decimal(row['quantity'])
      )
  );
}

export function loadProduction(filePath: string): ProductionVolume[] {
  const rows = readCsvFile(filePath);
  return rows.map(
    (row) =>
      new ProductionVolume(
        row['producer'],
        parseTimestamp(row['timestamp']),
        Duration.HOUR,
        new Decimal(row['quantity'])
      )
  );
}

export function loadAllocations(filePath: string): void {
  const rows = readCsvFile(filePath);
  for (const row of rows) {
    const producer = row['producer'];
    const consumer = row['consumer'];
    const periodStart = parseTimestamp(row['timestamp']);
    const duration = Duration.HOUR;
    const quantityKwh = new Decimal(row['quantity']);

    // TODO: return allocations
  }
}

export function getRepoRoot(currentDir?: string): string {
  if (!currentDir) {
    currentDir = __dirname;
  }

  const gitDir = path.join(currentDir, '.git');
  if (fs.existsSync(gitDir) && fs.statSync(gitDir).isDirectory()) {
    return currentDir;
  }

  const parentDir = path.dirname(currentDir);
  if (parentDir === currentDir) {
    throw new Error('Repository root not found');
  }

  return getRepoRoot(parentDir);
}
